src/utils/allocators.zig

```diff
@@ -136,38 +136,70 @@ pub fn RecycleFBA(config: struct {
     };
 }
 
+test "recycle allocator" {
+    const backing_allocator = std.testing.allocator;
+    var allocator = try RecycleFBA(.{}).init(backing_allocator, 1024);
+    defer allocator.deinit();
+
+    // alloc a slice of 100 bytes
+    const bytes = try allocator.allocator().alloc(u8, 100);
+    const ptr = bytes.ptr;
+    // free the slice
+    allocator.allocator().free(bytes);
+
+    // realloc should be the same (ie, recycled data)
+    const bytes2 = try allocator.allocator().alloc(u8, 100);
+    try std.testing.expectEqual(ptr, bytes2.ptr);
+    allocator.allocator().free(bytes2);
+
+    // same result with smaller slice
+    const bytes3 = try allocator.allocator().alloc(u8, 50);
+    try std.testing.expectEqual(ptr, bytes3.ptr);
+    allocator.allocator().free(bytes3);
+
+    // diff result with larger slice
+    const bytes4 = try allocator.allocator().alloc(u8, 200);
+    try std.testing.expect(ptr != bytes4.ptr);
+    allocator.allocator().free(bytes4);
+}
+
 /// thread safe disk memory allocator
 pub const DiskMemoryAllocator = struct {
-    filepath: []const u8,
-    count: std.atomic.Value(u64) = std.atomic.Value(u64).init(0),
-
+    dir: std.fs.Dir,
+    logger: sig.trace.Logger,
+    /// The address space mmap'd to a particular file will be at least
+    /// `(file_size * (1000 + mmap_ratio)) / 1000` in integer terms.
+    /// With a value of 0, the mmap'd size will be equal to the file size.
+    /// With a value of 1000, the mmap'd size will be double the file size.
+    /// NOTE: not intended to be changed after initialization.
+    mmap_ratio: MmapRatio = 0,
```

Comment:
I think protecting against the 0 footgun is more important; if it's misunderstood, the worst that can happen is that you end up mmapping a bit more than expected, within a relatively low magnitude.
If you think it's confusing, we can rename it to something more descriptive, ie `mmap_ratio_exp_offset`.
Mostly I allowed for thousands of decimal precision to facilitate a granular level of control. I can't predict what range of use cases one may have, this just allows us a bit of fine tuning, and costs relatively little to have - having it be a simple multiple was fairly limiting in the sizes that could be expressed; having the mmap size be more than double or triple the file size just doesn't seem like it would be very common, compared to wanting just a slightly larger margin for resizes.

Reply:
if we dont use it, it shouldnt be in the codebase because it will just make things more confusing - would rather remove and will be easy to add back if we need it

Reply:
I don't have a strong opinion about this being configurable. But I would not use a ratio of 1 (i.e. mmap_ratio = 0). With a ratio of 1, the resize function would only return true if you're resizing to a size that's within 1 page of the original allocation.

I don't think there's much cost of making the number very large. I would probably use a ratio of at least 8. 1024 or larger may also be fine. (i.e. mmap_ratio = 7000 or 1023000)

Reply:
Well it would be good to have a conclusive decision on this: I'm fine with:
* Status quo: no resize is supported.
* Current change: configurable ratio to allow a proportional range of resizability - perhaps better documented and simplified.
* Possible change: hard-code a ratio which we assume to be logical for our use cases.

If you have anything specific in mind, lmk, otherwise I'll just go with the first route, maintaining status quo with no resize support.

Reply:
> - Current change: configurable ratio to allow a proportional range of resizability - perhaps better documented and simplified.
> - Possible change: hard-code a ratio which we assume to be logical for our use cases.

I think we should go with one of these. Resize is useful, and you already have an implementation that looks good to me. Among these options, I'm fine deferring to either of your preferences @InKryption  and @0xNineteen.

If I have to make a choice, I would select simplified configurability, as in 
`mmap_size == mmap_ratio * file_size`. For me, this is the simplest/easiest to understand. But I see the advantages to other approaches too, so I'm not going to demand a specific approach here.

Reply:
`mmap_ratio * file_size` with a default ratio of 1 sounds good to me

Reply:
fc14c0746a8dca4c94fac28ab548c7a4795a3df0